File: packages/elixir/lib/janitor/protection.ex
defmodule Janitor.Protection do
  @moduledoc "Database-backed measurement limits and evaluator admission. Never an account authorization policy."
  alias Janitor.Storage, as: S

  def configure(nil), do: nil

  def configure(opts) do
    Janitor.Identity.validate_options!(opts)
    allowed!(opts, [:secret, :namespace, :requests, :evaluator, :on_event])

    requests =
      Keyword.merge(
        [global: 600, account: 60, session: 30, shards: 1, window_ms: 60_000],
        Keyword.get(opts, :requests, [])
      )

    evaluator =
      Keyword.merge(
        [
          max_calls: 120,
          window_ms: 60_000,
          max_concurrent: 4,
          failure_threshold: 3,
          cooldown_ms: 30_000
        ],
        Keyword.get(opts, :evaluator, [])
      )

    allowed!(requests, [:global, :account, :session, :shards, :window_ms])

    allowed!(evaluator, [
      :max_calls,
      :window_ms,
      :max_concurrent,
      :failure_threshold,
      :cooldown_ms
    ])

    for {k, v} <- requests ++ evaluator do
      {min, max} =
        cond do
          k in [:window_ms, :cooldown_ms] -> {1000, 3_600_000}
          k == :max_concurrent -> {1, 32}
          k == :shards -> {1, 128}
          k == :failure_threshold -> {1, 100}
          true -> {1, 1_000_000}
        end

      unless is_integer(v) and v >= min and v <= max,
        do: raise(ArgumentError, "invalid protection limit")
    end

    if opts[:on_event] && not is_function(opts[:on_event], 1),
      do: raise(ArgumentError, "invalid protection callback")

    opts
    |> Keyword.put(:requests, Map.new(requests))
    |> Keyword.put(:evaluator, Map.new(evaluator))
  end

  def admit(%{protection: nil}, _), do: :ok

  def admit(c, context) do
    context = context || %{}

    unless is_map(context) and
             Enum.all?(context, fn {k, v} ->
               k in [:account, :session] and is_binary(v) and byte_size(v) in 1..512
             end),
           do: raise(ArgumentError, "invalid admission context")

    limits = c.protection[:requests]
    now = Janitor.now()
    shards = min(limits.shards, limits.global)
    shard = :rand.uniform(shards) - 1

    global =
      if shards == 1 do
        {"global", limits.global, now}
      else
        {Jason.encode!(["global-shard-v1", shards, shard]),
         div(limits.global, shards) + if(shard < rem(limits.global, shards), do: 1, else: 0),
         div(now, limits.window_ms) * limits.window_ms}
      end

    checks =
      [global] ++
        for(
          k <- [:account, :session],
          context[k],
          do: {Jason.encode!([to_string(k), context[k]]), limits[k], now}
        )

    if Enum.all?(checks, fn {value, limit, window_time} ->
         S.query(
           c,
           "INSERT INTO #{S.table(c, "protection_quotas")} AS q (id,used,reset_at) VALUES ($1,1,$2) ON CONFLICT (id) DO UPDATE SET used = CASE WHEN q.reset_at <= $3 THEN 1 ELSE q.used + 1 END, reset_at = CASE WHEN q.reset_at <= $3 THEN $2 ELSE q.reset_at END WHERE q.reset_at <= $3 OR q.used < $4 RETURNING id",
           [key(c, value), window_time + limits.window_ms, window_time, limit]
         ) != []
       end),
       do: :ok,
       else:
         (
           emit(c, :request, :request_limit)
           {:error, {:rate_limited, ceil(limits.window_ms / 1000)}}
         )
  end

  def evaluate(%{protection: nil} = c, fun), do: Janitor.Bounded.run(fun, c.evaluator_timeout_ms)

  def evaluate(c, fun) do
    limits = c.protection[:evaluator]

    reservation =
      change(c, fn s, now ->
        s = Map.update!(s, "leases", &Enum.filter(&1, fn l -> l["expiresAt"] > now end))

        s =
          if now - s["windowStart"] >= limits.window_ms,
            do: %{s | "windowStart" => now, "used" => 0},
            else: s

        probe = s["openUntil"] != 0

        cond do
          s["openUntil"] > now ->
            {:return, {:denied, :circuit_open}}

          s["used"] >= limits.max_calls ->
            {:return, {:denied, :budget}}

          length(s["leases"]) >= limits.max_concurrent or
              (probe and Enum.any?(s["leases"], & &1["probe"])) ->
            {:return, {:denied, :concurrency}}

          true ->
            lease = %{
              "id" => Janitor.random_id("lease_"),
              "expiresAt" => now + c.evaluator_timeout_ms + 5000,
              "epoch" => s["epoch"],
              "probe" => probe
            }

            {:save, %{s | "leases" => s["leases"] ++ [lease], "used" => s["used"] + 1},
             {:lease, lease}}
        end
      end)

    case reservation do
      {:denied, reason} ->
        emit(c, :evaluator, reason)
        :unavailable

      {:lease, lease} ->
        result = Janitor.Bounded.run(fun, c.evaluator_timeout_ms, true)

        reason =
          case result do
            {:ok, value} -> if Janitor.Engine.valid_evaluation?(value), do: nil, else: :malformed
            :timeout -> :timeout
            _ -> :provider
          end

        if reason, do: emit(c, :evaluator, reason)
        finish(c, lease, reason)
        if reason, do: :unavailable, else: result
    end
  rescue
    _ ->
      emit(c, :evaluator, :storage)
      :unavailable
  catch
    _, _ ->
      emit(c, :evaluator, :storage)
      :unavailable
  end

  defp finish(c, lease, failure) do
    change(c, fn s, now ->
      if not Enum.any?(s["leases"], &(&1["id"] == lease["id"])) do
        {:return, :ok}
      else
        s =
          if failure == :timeout,
            do: s,
            else: Map.update!(s, "leases", &Enum.reject(&1, fn l -> l["id"] == lease["id"] end))

        s =
          cond do
            lease["epoch"] != s["epoch"] ->
              s

            is_nil(failure) ->
              s
              |> Map.put("failures", 0)
              |> then(fn s ->
                if lease["probe"], do: %{s | "openUntil" => 0, "epoch" => s["epoch"] + 1}, else: s
              end)

            lease["probe"] or s["failures"] + 1 >= c.protection[:evaluator].failure_threshold ->
              %{
                s
                | "openUntil" => now + c.protection[:evaluator].cooldown_ms,
                  "epoch" => s["epoch"] + 1
              }

            true ->
              Map.update!(s, "failures", &(&1 + 1))
          end

        {:save, s, :ok}
      end
    end)
  rescue
    _ -> emit(c, :evaluator, :storage)
  end

  defp change(c, fun, retries \\ 8)
  defp change(_, _, 0), do: raise("protection contention")

  defp change(c, fun, retries) do
    now = Janitor.now()
    id = key(c, "evaluator")

    row =
      List.first(
        S.query(c, "SELECT record FROM #{S.table(c, "evaluation_controls")} WHERE id = $1", [id])
      )

    state =
      if row,
        do: row["record"],
        else: %{
          "version" => 0,
          "windowStart" => now,
          "used" => 0,
          "failures" => 0,
          "openUntil" => 0,
          "epoch" => 0,
          "leases" => []
        }

    case fun.(state, now) do
      {:return, result} ->
        result

      {:save, next, result} ->
        next = Map.put(next, "version", state["version"] + 1)
        limits = c.protection[:evaluator]

        expiry =
          now +
            Enum.max([limits.window_ms, limits.cooldown_ms, c.evaluator_timeout_ms + 5000]) * 2

        rows =
          if state["version"] == 0 do
            S.query(
              c,
              "INSERT INTO #{S.table(c, "evaluation_controls")} (id,version,expires_at,record) VALUES ($1,$2,$3,$4) ON CONFLICT (id) DO NOTHING RETURNING id",
              [id, next["version"], expiry, next]
            )
          else
            S.query(
              c,
              "UPDATE #{S.table(c, "evaluation_controls")} SET version=$2,expires_at=$3,record=$4 WHERE id=$1 AND version=$5 RETURNING id",
              [id, next["version"], expiry, next, state["version"]]
            )
          end

        if rows == [], do: change(c, fun, retries - 1), else: result
    end
  end

  def cleanup(c) do
    for {table, field} <- [
          {"protection_quotas", "reset_at"},
          {"evaluation_controls", "expires_at"}
        ] do
      S.query(
        c,
        "DELETE FROM #{S.table(c, table)} WHERE id IN (SELECT id FROM #{S.table(c, table)} WHERE #{field} <= $1 ORDER BY #{field} LIMIT 100) AND #{field} <= $1",
        [Janitor.now()]
      )
    end
  end

  defp key(c, value),
    do: Janitor.Identity.label(c.protection, Jason.encode!(["protection-v1", value]))

  defp allowed!(opts, keys),
    do:
      if(not Keyword.keyword?(opts) or Enum.any?(Keyword.keys(opts), &(&1 not in keys)),
        do: raise(ArgumentError, "unknown protection option")
      )

  defp emit(c, kind, reason) do
    if fun = c.protection[:on_event], do: fun.(%{kind: kind, reason: reason})
  rescue
    _ -> :ok
  catch
    _, _ -> :ok
  end
end

File: packages/adapters/src/protection.ts
import { z } from "zod";
import {
  isEvaluation,
  isLookupScope,
  isCandidateEvaluations,
  isCrossDevicePrediction,
  isApiActivityRisk,
  isOperatorEvaluation,
} from "@aarondovturkel/doorman-core";
import type {
  EvaluationControl,
  EvaluationLease,
  ProtectionStorage,
  VisitorEvaluator,
} from "@aarondovturkel/doorman-core";
import { createSubjectLinker } from "./subject.js";
import type { SubjectLinkingOptions } from "./subject.js";

const limit = z.number().int().min(1).max(1_000_000);
const windowMs = z.number().int().min(1000).max(3_600_000);
const config = z
  .object({
    requests: z
      .object({
        global: limit.default(600),
        account: limit.default(60),
        session: limit.default(30),
        shards: z.number().int().min(1).max(128).default(1),
        windowMs: windowMs.default(60_000),
      })
      .strict()
      .default({
        global: 600,
        account: 60,
        session: 30,
        shards: 1,
        windowMs: 60_000,
      }),
    evaluator: z
      .object({
        maxCalls: limit.default(120),
        windowMs: windowMs.default(60_000),
        maxConcurrent: z.number().int().min(1).max(32).default(4),
        failureThreshold: z.number().int().min(1).max(100).default(3),
        cooldownMs: windowMs.default(30_000),
      })
      .strict()
      .default({
        maxCalls: 120,
        windowMs: 60_000,
        maxConcurrent: 4,
        failureThreshold: 3,
        cooldownMs: 30_000,
      }),
  })
  .strict();
export type ProtectionReason =
  | "request-limit"
  | "budget"
  | "concurrency"
  | "circuit-open"
  | "timeout"
  | "provider"
  | "malformed"
  | "storage"
  | "contention";
export type ProtectionOptions = SubjectLinkingOptions &
  z.input<typeof config> & {
    onEvent?: (event: {
      kind: "request" | "evaluator";
      reason: ProtectionReason;
    }) => void;
  };
export type AdmissionContext = { account?: string; session?: string };
const admission = z
  .object({
    account: z.string().min(1).max(512).optional(),
    session: z.string().min(1).max(512).optional(),
  })
  .strict();
const RETRIES = 8;

/** No raw account/session keys are persisted. Configure the same namespace on every replica. */
export function createProtection(
  storage: ProtectionStorage,
  options: ProtectionOptions,
  timeoutMs: number,
) {
  if (!Number.isInteger(timeoutMs) || timeoutMs < 1 || timeoutMs > 5000)
    throw new Error("Protected evaluation timeout must be 1–5000ms");
  const limits = config.parse({
    requests: options.requests,
    evaluator: options.evaluator,
  });
  const label = createSubjectLinker(options);
  const key = (value: string) =>
    label(JSON.stringify(["protection-v1", value]));
  const emit = (kind: "request" | "evaluator", reason: ProtectionReason) => {
    try {
      options.onEvent?.({ kind, reason });
    } catch {
      /* Metrics cannot change policy. */
    }
  };
  const controlKey = key("evaluator");
  const shards = Math.min(limits.requests.shards, limits.requests.global);
  let nextShard = crypto.getRandomValues(new Uint32Array(1))[0]! % shards;
  const ttl =
    Math.max(
      limits.evaluator.windowMs,
      limits.evaluator.cooldownMs,
      timeoutMs + 5000,
    ) * 2;
  async function change<T>(
    fn: (
      state: EvaluationControl,
      now: number,
    ) => { state?: EvaluationControl; result: T },
  ): Promise<T> {
    for (let attempt = 0; attempt < RETRIES; attempt++) {
      const now = Date.now();
      const previous = await storage.getControl(await controlKey);
      const state: EvaluationControl = previous ?? {
        version: 0,
        windowStart: now,
        used: 0,
        failures: 0,
        openUntil: 0,
        epoch: 0,
        leases: [],
      };
      const next = fn(structuredClone(state), now);
      if (!next.state) return next.result;
      next.state.version = state.version + 1;
      if (
        await storage.compareControl(
          await controlKey,
          state.version,
          next.state,
          now + ttl,
        )
      )
        return next.result;
    }
    throw new Error("Protection contention");
  }
  return {
    async admit(context: AdmissionContext = {}) {
      const parsed = admission.parse(context);
      const now = Date.now();
      const shard = nextShard;
      nextShard = (nextShard + 1) % shards;
      // Static slices sum to the configured global maximum. No borrowing or retry.
      // Epoch-aligned windows keep every slice on the same clock boundary.
      const checks: [string, number, number][] = [
        shards === 1
          ? ["global", limits.requests.global, now]
          : [
              JSON.stringify(["global-shard-v1", shards, shard]),
              Math.floor(limits.requests.global / shards) +
                Number(shard < limits.requests.global % shards),
              Math.floor(now / limits.requests.windowMs) *
                limits.requests.windowMs,
            ],
      ];
      if (parsed.account)
        checks.push([
          JSON.stringify(["account", parsed.account]),
          limits.requests.account,
          now,
        ]);
      if (parsed.session)
        checks.push([
          JSON.stringify(["session", parsed.session]),
          limits.requests.session,
          now,
        ]);
      // Earlier quotas count denied attempts too. Each UPSERT is atomic across replicas.
      for (const [value, maximum, windowTime] of checks) {
        if (
          !(await storage.consumeQuota(
            await key(value),
            maximum,
            limits.requests.windowMs,
            windowTime,
          ))
        ) {
          emit("request", "request-limit");
          return {
            allowed: false,
            retryAfterSeconds: Math.ceil(limits.requests.windowMs / 1000),
          };
        }
      }
      return { allowed: true, retryAfterSeconds: 0 };
    },
    wrap(evaluator: VisitorEvaluator): VisitorEvaluator {
      async function guarded<T>(
        run: () => Promise<T>,
        validate: (value: unknown) => value is T,
        method: keyof NonNullable<VisitorEvaluator["requestCosts"]>,
      ): Promise<T> {
        const cost = evaluator.requestCosts?.[method] ?? 1;
        if (!Number.isInteger(cost) || cost < 1 || cost > 16)
          throw new Error("Invalid evaluator request cost");
        let reservation: {
          lease?: EvaluationLease;
          reason?: ProtectionReason;
        };
        try {
          reservation = await change<{
            lease?: EvaluationLease;
            reason?: ProtectionReason;
          }>((s, now) => {
            s.leases = s.leases.filter((l) => l.expiresAt > now);
            if (s.openUntil > now)
              return { result: { reason: "circuit-open" as const } };
            if (now - s.windowStart >= limits.evaluator.windowMs) {
              s.windowStart = now;
              s.used = 0;
            }
            if (s.used + cost > limits.evaluator.maxCalls)
              return { result: { reason: "budget" as const } };
            const probe = s.openUntil !== 0;
            if (
              s.leases.reduce((sum, l) => sum + (l.cost ?? 1), 0) + cost >
                limits.evaluator.maxConcurrent ||
              (probe && s.leases.some((l) => l.probe))
            )
              return { result: { reason: "concurrency" as const } };
            const lease = {
              id: crypto.randomUUID(),
              expiresAt: now + timeoutMs + 5000,
              epoch: s.epoch,
              probe,
              cost,
            };
            s.leases.push(lease);
            s.used += cost;
            return { state: s, result: { lease } };
          });
        } catch {
          emit("evaluator", "storage");
          throw new Error("Evaluation protection unavailable");
        }
        const { lease, reason } = reservation;
        if (!lease) {
          emit("evaluator", reason!);
          throw new Error("Evaluation admission denied");
        }
        let failure: ProtectionReason | undefined;
        let timer: ReturnType<typeof setTimeout> | undefined;
        try {
          const value = await Promise.race([
            Promise.resolve().then(() => run()),
            new Promise<never>((_, reject) => {
              timer = setTimeout(() => {
                failure = "timeout";
                reject(new Error("Evaluation timeout"));
              }, timeoutMs);
            }),
          ]);
          if (!validate(value)) {
            failure = "malformed";
            throw new Error("Malformed evaluation");
          }
          return value;
        } catch (error) {
          failure ??= "provider";
          emit("evaluator", failure);
          throw error;
        } finally {
          clearTimeout(timer);
          try {
            await change((s, now) => {
              if (!s.leases.some((l) => l.id === lease.id))
                return { result: undefined };
              // A timed-out provider may still be running. Keep its lease until expiry.
              if (failure !== "timeout")
                s.leases = s.leases.filter((l) => l.id !== lease.id);
              // Old completions cannot close a newer breaker or release another lease.
              if (lease.epoch === s.epoch) {
                if (!failure) {
                  s.failures = 0;
                  if (lease.probe) {
                    s.openUntil = 0;
                    s.epoch++;
                  }
                } else if (
                  lease.probe ||
                  ++s.failures >= limits.evaluator.failureThreshold
                ) {
                  s.openUntil = now + limits.evaluator.cooldownMs;
                  s.epoch++;
                }
              }
              return { state: s, result: undefined };
            });
          } catch {
            emit(
              "evaluator",
              "storage",
            ); /* Abandoned leases expire; never retry inference. */
          }
        }
      }
      return {
        requestCosts: evaluator.requestCosts,
        ...(evaluator.evaluateOperator
          ? {
              evaluateOperator: (
                input: Parameters<
                  NonNullable<VisitorEvaluator["evaluateOperator"]>
                >[0],
              ) =>
                guarded(
                  () => evaluator.evaluateOperator!(input),
                  isOperatorEvaluation,
                  "evaluateOperator",
                ),
            }
          : {}),
        ...(evaluator.evaluateActivity
          ? {
              evaluateActivity: (
                input: Parameters<
                  NonNullable<VisitorEvaluator["evaluateActivity"]>
                >[0],
              ) =>
                guarded(
                  () => evaluator.evaluateActivity!(input),
                  isApiActivityRisk,
                  "evaluateActivity",
                ),
            }
          : {}),
        evaluate: (input) =>
          guarded(() => evaluator.evaluate(input), isEvaluation, "evaluate"),
        ...(evaluator.planLookup
          ? {
              planLookup: (
                input: Parameters<
                  NonNullable<VisitorEvaluator["planLookup"]>
                >[0],
              ) =>
                guarded(
                  () => evaluator.planLookup!(input),
                  isLookupScope,
                  "planLookup",
                ),
            }
          : {}),
        ...(evaluator.evaluateCandidates
          ? {
              evaluateCandidates: (
                input: Parameters<
                  NonNullable<VisitorEvaluator["evaluateCandidates"]>
                >[0],
              ) =>
                guarded(
                  () => evaluator.evaluateCandidates!(input),
                  isCandidateEvaluations,
                  "evaluateCandidates",
                ),
            }
          : {}),
        ...(evaluator.predictIdentity
          ? {
              predictIdentity: (
                input: Parameters<
                  NonNullable<VisitorEvaluator["predictIdentity"]>
                >[0],
              ) =>
                guarded(
                  () => evaluator.predictIdentity!(input),
                  isCrossDevicePrediction,
                  "predictIdentity",
                ),
            }
          : {}),
      };
    },
    cleanup: () => storage.cleanupProtection(Date.now(), 100),
  };
}
